src/ucp/proto/proto_single.inl

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCP_PROTO_SINGLE_INL_
+#define UCP_PROTO_SINGLE_INL_
+
+#include "proto_single.h"
+#include "proto_common.inl"
+
+#include <ucp/dt/datatype_iter.inl>
+
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_proto_am_bcopy_single_send(ucp_request_t *req, ucp_am_id_t am_id,
+                               ucp_lane_index_t lane,
+                               uct_pack_callback_t pack_func, void *pack_arg,
+                               size_t max_packed_size)
+{
+    ucp_ep_t *ep                       = req->send.ep;
+    const uct_iface_attr_t *iface_attr = ucp_ep_get_iface_attr(ep, lane);
+    ssize_t packed_size;
+    uint64_t *buffer;
+
+    if ((max_packed_size <= UCS_ALLOCA_MAX_SIZE) &&
+        (iface_attr->cap.flags & UCT_IFACE_FLAG_AM_SHORT) &&
+        (max_packed_size <= iface_attr->cap.am.max_short)) {
+
+        /* Send as inline if expected size is small enough */
+        buffer      = ucs_alloca(max_packed_size);
+        packed_size = pack_func(buffer, pack_arg);
+        ucs_assertv((packed_size >= 0) && (packed_size <= max_packed_size),
+                    "packed_size=%zd max_packed_size=%zu", packed_size,
+                    max_packed_size);
+        return uct_ep_am_short(ep->uct_eps[lane], am_id, buffer[0],
+                               &buffer[1], packed_size - sizeof(buffer[0]));
+    } else {
+        /* Send as bcopy */
+        packed_size = uct_ep_am_bcopy(ep->uct_eps[lane], am_id, pack_func,
+                                      pack_arg, 0);
+        return ucs_likely(packed_size >= 0) ? UCS_OK : packed_size;
+    }
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_proto_am_bcopy_single_progress(ucp_request_t *req, ucp_am_id_t am_id,
+                                   ucp_lane_index_t lane,
+                                   uct_pack_callback_t pack_func, void *pack_arg,
+                                   size_t max_packed_size,
+                                   ucp_proto_complete_cb_t complete_func,
+                                   ucp_proto_complete_cb_t error_func)
+{
+    ucs_status_t status;
+
+    ucs_assert(error_func != NULL);
+
+    status = ucp_proto_am_bcopy_single_send(req, am_id, lane, pack_func,
+                                            pack_arg, max_packed_size);
+    if (ucs_likely(status == UCS_OK)) {
+        if (complete_func != NULL) {
```

Comment:
it is always not a `NULL` (in the `ucp_eager_bcopy_single_progress()`), or I don't see other usages?

Reply:
it's in code which will be in later PR